src/batch.rs

```diff
@@ -0,0 +1,61 @@
+//! # Batch
+//!
+//! This module contains types for batch operations. A batch operation is a
+//! collection of write operations (puts and/or deletes) that are applied
+//! atomically to the database.
+
+/// A batch of write operations (puts and/or deletes). All operations in the
+/// batch are applied atomically to the database. If multiple operations appear
+/// for a a single key, the last operation will be applied. The others will be
+/// dropped.
+///
+/// # Examples
+/// ```rust,no_run,compile_fail
+/// use slatedb::batch::{WriteBatch, WriteOp};
+///
+/// let object_store = Arc::new(LocalFileSystem::new());
+/// let db = Db::open("path/to/db".into(), object_store).await;
+///
+/// let mut batch = WriteBatch::new();
```

Comment:
how do we want timestamp semantics to work for a batch? should they all use the same timestamp at the time the batch was written? alternatively we can use the clock at the time the record was added to the batch (I think this works better for long-running batches)

Reply:
> should they all use the same timestamp at the time the batch was written?

This would likely result in records entering the database with out-of-order timestamps, which technically can't happen with regular inserts because the clock advances monotonically. I don't think that's problematic from a TTL perspective, but it does mean that the db wouldn't respect timestamp-order during retrieval.

Reply:
in the sense of making Snapshot with sequence number, the keys in the write batch are expected to have the same sequence number on commits.

Reply:
> This would likely result in records entering the database with out-of-order timestamps

do we support inserting other records while a write batch is being created? that seems fragile since you can't guarantee that the same key won't be written to outside of the write batch (almost like having multiple writers, right?)

if that's not the scenario you're thinking of what are you thinking about that would cause out-of-order timestamps?

Reply:
That is the scenario I had in mind. I was thinking that to achieve that kind of isolation you'd use the transaction stuff that @flaneur2020 is sketching out.

Reply:
I assumed we would allow inserts to happen while we were creating a batch. One change I've made today is centralizing the `write()`'s on to a single async/await loop in a writer thread. This should serialize all writes. If we do the clock() call in the [`write_batch_request`](https://github.com/slatedb/slatedb/pull/264/files#diff-51529babecd6e4018b78a49622df5b5d2c480476389ec07c0b4943d8bdbaa35eR24), I think it should be OK?

Reply:
> This would likely result in records entering the database with out-of-order timestamps, which technically can't happen with regular inserts because the clock advances monotonically. I don't think that's problematic from a TTL perspective, but it does mean that the db wouldn't respect timestamp-order during retrieval.

@rodesai I don't quite follow what you're saying here. Maybe an example might help me.. 🧠

Reply:
> That is the scenario I had in mind. I was thinking that to achieve that kind of isolation you'd use the transaction stuff that @flaneur2020 is sketching out.

👍 Use transactions (#248) if you want to keep other writes from stomping on the batch.

Thinking a bit more, even without transactions, this seems like the pattern we'd have:

```
time0 thread1 create_batch
time1 thread1 batch.put(k1, v1)
time2 thread2 db.put(k1, v2)
time3 thread2 db.put(k2 v1)
time4 thread1 db.write(batch)
```

In this sequence, we'd end up with (k1, v1, t4) and (k2, v1, t3). That seems.. right? Are there other examples that cause problems?

Reply:
> @rodesai I don't quite follow what you're saying here. Maybe an example might help me.. 🧠

I quoted the wrong part in my reply. The out-of-order (with respect to timestamp) inserts would occur if the timestamp is determined at the time a record is added to the batch. If we use the timestamp the batch is inserted to the db then this can't happen.

Reply:
😅 That's what I was hoping you'd say. The quote didn't seem to match. 😆  Are you good with using the batch timestamp? It seems more intuitive to me _except_ when using custom clocks. But I think that's a good tradeoff? For a custom clock you'd have to use the max of the records in the batch... 🤔 @agavra @vigneshc WDYT about this?

Reply:
[TTL RFC](https://github.com/slatedb/slatedb/blob/main/rfcs/0003-timestamps-and-ttl.md) only lets us provide the clock, and db will call `now()` when appropriate. With that, we cannot control the timestamp of each record really. If we read a batch of records from a log, and insert it to SlateDB, all of them will have the same timestamp (even without batch).

I think having a single timestamp for all the records, retrieved when `WriteBatch` call is called seems more intuitive to me . They all appear in an instant, atomically. So they must have the same timestamp.